pkg/provider/aws/resources/eks.go

```diff
@@ -65,43 +110,122 @@ type (
+func CreateEksCluster(cfg *config.Application, clusterName string, subnets []*Subnet, securityGroups []*any, units []*core.ExecutionUnit, dag *core.ResourceGraph) {
 	references := []core.AnnotationKey{}
 	for _, u := range units {
 		references = append(references, u.Provenance())
 	}
 
+	appName := cfg.AppName
+
 	clusterRole := createClusterAdminRole(appName, clusterName+"-k8sAdmin", references)
 	dag.AddResource(clusterRole)
 
 	cluster := NewEksCluster(appName, clusterName, subnets, securityGroups, clusterRole)
 	cluster.ConstructsRef = references
-	dag.AddResource(cluster)
-	dag.AddDependency(cluster, clusterRole)
+	dag.AddDependenciesReflect(cluster)
+
+	createNodeGroups(cfg, dag, units, clusterName, cluster, subnets)
 
 	fargateRole := createPodExecutionRole(appName, clusterName+"-FargateExecutionRole", references)
-	dag.AddResource(fargateRole)
+	dag.AddDependenciesReflect(fargateRole)
 
 	profile := NewEksFargateProfile(cluster, subnets, fargateRole, references)
 	profile.Selectors = append(profile.Selectors, &FargateProfileSelector{Namespace: "default", Labels: map[string]string{"klotho-fargate-enabled": "true"}})
+	dag.AddDependenciesReflect(profile)
+}
 
-	dag.AddResource(profile)
-	dag.AddDependency(profile, fargateRole)
-	dag.AddDependency(profile, cluster)
+func createNodeGroups(cfg *config.Application, dag *core.ResourceGraph, units []*core.ExecutionUnit, clusterName string, cluster *EksCluster, subnets []*Subnet) []*EksNodeGroup {
+	type groupKey struct {
+		InstanceType string
+		NetworkType  string
+	}
 
-	nodeRole := createNodeRole(appName, clusterName+"-NodeGroupRole", references)
-	dag.AddResource(nodeRole)
+	type groupSpec struct {
+		DiskSizeGiB int
+		refs        []core.AnnotationKey
+	}
 
-	nodeGroup := NewEksNodeGroup(cluster, subnets, nodeRole, references)
-	dag.AddResource(nodeGroup)
-	dag.AddDependency(nodeGroup, nodeRole)
-	dag.AddDependency(nodeGroup, cluster)
+	groupSpecs := make(map[groupKey]*groupSpec)
+
+	for _, unit := range units {
+		unitCfg := cfg.GetExecutionUnit(unit.ID)
+		params := unitCfg.GetExecutionUnitParamsAsKubernetes()
+		key := groupKey{InstanceType: params.InstanceType, NetworkType: unitCfg.NetworkPlacement}
+		spec := groupSpecs[key]
+		if spec == nil {
+			spec = &groupSpec{
+				DiskSizeGiB: 20,
+			}
+			groupSpecs[key] = spec
+		}
+		spec.refs = append(spec.refs, unit.AnnotationKey)
+		if params.DiskSizeGiB > spec.DiskSizeGiB {
+			spec.DiskSizeGiB = params.DiskSizeGiB
+		}
+	}
+
+	var groups []*EksNodeGroup
+
+	hasInstanceType := false
+	for gk := range groupSpecs {
+		if gk.InstanceType != "" {
+			hasInstanceType = true
+			break
+		}
+	}
+	if !hasInstanceType {
+		for gk, spec := range groupSpecs {
+			if gk.InstanceType == "" {
+				groupSpecs[groupKey{InstanceType: "t3.medium", NetworkType: gk.NetworkType}] = spec
+			}
+		}
+	}
```

Comment:
this will still leave blank instance types in the group specs list, which i believe would cause failures if there is another node grup thats set

Reply:
> this will still leave blank instance types in the group specs list, which i believe would cause failures if there is another node grup thats set

It shouldn't due to the `continue` here:
```go
	for groupKey, spec := range groupSpecs {
		if groupKey.InstanceType == "" {
			continue
		}
```
I could delete the key instead if you think that's clearer.

> might want to split this logic out so it can be unit tested, i think thatll help us understand what will get created

Split out more than `createNodeGroups` (with tests in https://github.com/klothoplatform/klotho/pull/487/files#diff-03d25bd6b71fc8fcc391be68236b07406fbb298a04ada456b9f28794e7c50b9aR143)?